### PR TITLE
Adding function for collecting all expressions with unit inconsistencies

### LINKED ISSUE #2893: Request for an `identify_inconsistent_units` method
## Summary

The Pyomo unit consistency checks current have an `assert_units_consistent` function which will iterate over the constraints in a model and raise a `UnitsError` if a constraint is encountered with inconsistent units. Whilst this is useful for testing, in most cases a model developer would like to get a list of all constraints with unit consistency issues, not just an exception for the first issue encountered.

Additionally, the `assert_units_consistent` function has some `print` statements embedded in it which clutter the output somewhat which could probably be removed (the proposed `identify_inconsistent_units` method would provide a means for getting the full list of problematic constraints in its place).

### Rationale

I am writing a tool as part of https://github.com/IDAES/idaes-pse/issues/1208 which aims to provide users with a summary of all (or at least most) the issues in their model, one of which is unit consistency. As such I would like to generate the list of problematic constraints for the user to reference and to do so without any extraneous text output.

### Description

For the time being, I am using:

```
constraints_with_inconsistent_units = ComponentSet()
for c in self.model.component_data_objects(Constraint, descend_into=True):
    try:
        assert_units_consistent(c)
    except UnitsError:
        constraints_with_inconsistent_units.add(c)
return constraints_with_inconsistent_units
```

This generates the desired list, but produces a line of text for each failing constraint; e.g. `Error in units when checking b.c4`.

### PR DESCRIPTION
**Pull request by andrewlee94**
## Summary/Motivation:
This PR addresses the request for a function to return a list of all expressions in a model with unit consistency errors as well as cleaning up print statements in the existing `assert_units_consistent` function.

## Changes proposed in this PR:
- Remove print statements from `assert_units_consistent` function
- New function to iterate overall `Constraints`, `Expressions` and `Objectives` in a `Block` and return a `ComponentSet` of all components which fail unit consistency checks.

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/util/check_units.py [hunk 1]
```diff
@@ -261,14 +258,36 @@ def assert_units_consistent(obj):
...
-                print('Error in units when checking {}'.format(cdata))
-                raise
+            handler(cdata)
     else:
+        handler(obj)
+
+
+def identify_inconsistent_units(block):
+    """
+    This function generates a ComponentSet of all Constraints, Expression and Objectives
```
Comment (blnicho) [inline pyomo/util/check_units.py:268]: Typo/grammar: "...Constraints, Expressions, and Objectives..."

**Review by blnicho — APPROVED** [review]
I found one minor typo in a docstring but otherwise I think this looks fine.

**Review by mrmundt — APPROVED** [review]

File: pyomo/util/check_units.py [hunk 2]
```diff
@@ -240,11 +241,7 @@ def assert_units_consistent(obj):
     if objtype in native_types:
         return
     elif obj.is_expression_type() or objtype is IndexTemplate:
-        try:
-            _assert_units_consistent_expression(obj)
-        except UnitsError:
-            print('Units problem with expression {}'.format(obj))
```
Comment (jsiirola) [inline pyomo/util/check_units.py:246]: I agree with removing these print statements, but should the message still be sent out the logger?  (i.e., I don't know if the exception generated within the walker contains enough information - for example, does it include the constraint / objective name?

**Comment by andrewlee94** [comment]
@jsiirola @mrmundt I've just pushed changes to address your comments.

**Review by jsiirola — APPROVED** [review]

**Review by mrmundt — APPROVED** [review]

**Pull request merged**